### PR TITLE
Add configuration options to `App` to choose reconciler

### PR DESCRIPTION
**Pull request by carson-katri**
This PR enables support for `App` + `Scene` in the Fiber reconciler, and provides a way for apps to specify whether they want to use the `FiberReconciler`, and if they want to enable custom layout.

```swift
@main
struct MyApp: App {
  static let _configuration: _AppConfiguration = .init(
    reconciler: .fiber(useDynamicLayout: true)
  )
  
  var body: some Scene {
    WindowGroup {
      Text("Hello, world!")
    }
  }
}
```

### CODE REVIEW

**Comment by ezraberch** [comment]
I'd prefer this interface:

```swift
  static let _configuration: _AppConfiguration = .init(
    useReconciler: .fiber,
    shouldLayout: true
  )

```

**Comment by carson-katri** [comment]
I updated it to this:
```swift
static let _configuration: _AppConfiguration = .init(
  reconciler: .fiber(shouldLayout: true)
)
```
Reconciler options are:
```swift
.stack
.fiber()
.fiber(shouldLayout: true)
```
You can also pass the root environment in the configuration. Let me know what you think.

File: Sources/TokamakCore/App/App.swift [hunk 1]
```diff
@@ -28,22 +28,47 @@ public protocol App: _TitledApp {
...
+    reconciler: Reconciler = .stack,
+    rootEnvironment: EnvironmentValues = .init()
+  ) {
+    self.reconciler = reconciler
+    self.rootEnvironment = rootEnvironment
+  }
+
+  public enum Reconciler {
+    case stack
+    case fiber(shouldLayout: Bool = false)
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/App/App.swift:63]: WDYT about a slightly different naming here?

```suggestion
    case fiber(usesDynamicLayout: Bool = false)
```

File: Sources/TokamakCore/Fiber/FiberReconciler+TreeReducer.swift [hunk 2]
```diff
@@ -57,9 +57,61 @@ extension FiberReconciler {
...
+        },
+        visitChildren: { $0._visitChildren }
+      )
+      partialResult = result
+    }
+
     static func reduce<V>(into partialResult: inout Result, nextView: V) where V: View {
+      var result = partialResult
+      Self.reduce(
+        into: &result,
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/FiberReconciler+TreeReducer.swift:86]: Is there a need to copy to temporary `result` here? Can't `partialResult` be passed directly?

File: Sources/TokamakCore/Fiber/Scene/Fiber+Scene.swift [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Created by Carson Katri on 5/30/22.
+//
+
+@_spi(TokamakCore)
+public extension FiberReconciler.Fiber {}
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/Scene/Fiber+Scene.swift:19]: Is this supposed to stay empty? What's the purpose of it then?

File: Sources/TokamakCore/Fiber/Scene/SceneVisitor.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  static func reduce<S: Scene>(partialResult: Result, nextScene: S) -> Result {
+    var result = partialResult
+    Self.reduce(into: &result, nextScene: nextScene)
+    return result
+  }
+}
+
+final class SceneReducerVisitor<R: SceneReducer>: SceneVisitor {
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/Scene/SceneVisitor.swift:46]: Let's add some doc comments to types in this file.

File: Sources/TokamakCore/Fiber/FiberReconciler+TreeReducer.swift (hunk 2 above)
Comment (carson-katri) [reply Sources/TokamakCore/Fiber/FiberReconciler+TreeReducer.swift:86]: `createFiber` accesses it, and mutation requires exclusive access, so I have to make a copy AFAIK. But I could also rework `createFiber` to pass all the parameters in.

File: Sources/TokamakCore/Fiber/Scene/Fiber+Scene.swift (hunk 3 above)
Comment (carson-katri) [reply Sources/TokamakCore/Fiber/Scene/Fiber+Scene.swift:19]: I accidentally forgot to delete it 🙃

File: Sources/TokamakCore/App/App.swift (hunk 1 above)
Comment (carson-katri) [reply Sources/TokamakCore/App/App.swift:63]: `useDynamicLayout` sounds good to me, I'll also rename `shouldLayout` in the `FiberRenderer` protocol.

**Comment by carson-katri** [comment]
`App` should now properly handle `@State`.

File: Sources/TokamakCore/App/Scenes/Scene.swift [hunk 5]
```diff
@@ -21,8 +21,14 @@ public protocol Scene {
   // FIXME: If I put `@SceneBuilder` in front of this
   // it fails to build with no useful error message.
   var body: Self.Body { get }
+
+  func _visitChildren<V: SceneVisitor>(_ visitor: V)
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/App/Scenes/Scene.swift:25]: nit: would you mind adding doc comments to these two functions?

File: Sources/TokamakCore/Fiber/App/AppVisitor.swift [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Created by Carson Katri on 5/31/22.
+//
+
+/// A type that can visit a `App`.
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/App/AppVisitor.swift:18]: nit
```suggestion
/// A type that can visit an `App`.
```

File: Sources/TokamakCore/Fiber/Fiber.swift [hunk 7]
```diff
@@ -37,27 +37,23 @@ public extension FiberReconciler {
   /// After the entire tree has been traversed, the current and work in progress trees are swapped,
   /// making the updated tree the current one,
   /// and leaving the previous current tree available to apply future changes on.
-  final class Fiber: CustomDebugStringConvertible {
+  class Fiber {
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/Fiber.swift:40]: nit: Is `Fiber` supposed to be subclassed anywhere? If not, what's the purpose of `final` removal?

File: Sources/TokamakCore/Fiber/Fiber.swift [hunk 8]
```diff
@@ -253,46 +239,172 @@ public extension FiberReconciler {
...
+        // Create the alternate lazily
+        let alternate = Fiber(
+          bound: alternateApp,
+          alternate: self,
+          outputs: self.outputs,
+          typeInfo: self.typeInfo,
+          element: self.element,
+          reconciler: reconciler
+        )
+        self.alternate = alternate
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/Fiber.swift:289]: doesn't this capture `self` strongly and create a reference cycle?

File: Sources/TokamakCore/Fiber/Fiber.swift [hunk 9]
```diff
@@ -253,46 +239,172 @@ public extension FiberReconciler {
...
+          bound: alternateScene,
+          alternate: self,
+          outputs: self.outputs,
+          typeInfo: self.typeInfo,
+          element: self.element,
+          parent: self.parent?.alternate,
+          elementParent: self.elementParent?.alternate,
+          reconciler: reconciler
+        )
+        self.alternate = alternate
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/Fiber.swift:354]: similar comment about strong capture of `self` and reference cyclea

File: Sources/TokamakCore/Fiber/FiberReconciler.swift [hunk 10]
```diff
@@ -70,7 +70,31 @@ public final class FiberReconciler<Renderer: FiberRenderer> {
     reconcile(from: current)
   }
 
-  final class ReconcilerVisitor: ViewVisitor {
+  public init<A: App>(_ renderer: Renderer, _ app: A) {
+    self.renderer = renderer
+    var environment = renderer.defaultEnvironment
+    environment.measureText = renderer.measureText
+//    current = .init(
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Fiber/FiberReconciler.swift:77]: is this supposed to stay commented?

File: Sources/TokamakDOM/App/App.swift [hunk 11]
```diff
@@ -21,8 +21,15 @@ import TokamakCore
...
 
 public extension App {
-  static func _launch(_ app: Self, _ rootEnvironment: EnvironmentValues) {
-    _launch(app, rootEnvironment, TokamakDOM.body)
+  static func _launch(_ app: Self, with configuration: _AppConfiguration) {
+    switch configuration.reconciler {
+    case .stack:
+      _launch(app, configuration.rootEnvironment, TokamakDOM.body)
+    case let .fiber(useDynamicLayout):
+      _ = Unmanaged.passRetained(
```
Comment (MaxDesiatov) [inline Sources/TokamakDOM/App/App.swift:29]: could you clarify the use of `Unmanaged` here in a comment?

Comment (carson-katri) [reply Sources/TokamakDOM/App/App.swift]: Actually wasn't needed 👍

File: Sources/TokamakCore/Fiber/Fiber.swift (hunk 8 above)
Comment (carson-katri) [reply Sources/TokamakCore/Fiber/Fiber.swift:289]: I don't believe it does... `alternate` is a `weak` property, and its binding to a fresh `Fiber` instance with no strong reference to itself, only the parent (or the left sibling) of the alternate keeps a strong reference to it. Unless I'm overlooking something.

Comment (carson-katri) [reply Sources/TokamakCore/Fiber/Fiber.swift:291]: Oh, unless you meant the closure itself, not the `alternate` property. I think you're right, I just added `[weak self]` to it.

File: Sources/TokamakDOM/App/App.swift [hunk 12]
```diff
@@ -21,8 +21,15 @@ import TokamakCore
...
 
 public extension App {
-  static func _launch(_ app: Self, _ rootEnvironment: EnvironmentValues) {
-    _launch(app, rootEnvironment, TokamakDOM.body)
+  static func _launch(_ app: Self, with configuration: _AppConfiguration) {
+    switch configuration.reconciler {
+    case .stack:
+      _launch(app, configuration.rootEnvironment, TokamakDOM.body)
+    case let .fiber(useDynamicLayout):
+//      _ = Unmanaged.passRetained(
```
Comment (MaxDesiatov) [inline Sources/TokamakDOM/App/App.swift:29]: not sure if PR is ready for review, but if so, this commented code needs to be cleaned up

**Review by MaxDesiatov — APPROVED** [review]
Thank you for addressing all the nits! 😅 This looks great 👍

**Pull request merged**